puppet/zulip/manifests/supervisor.pp

```diff
@@ -1,16 +1,18 @@
 class zulip::supervisor {
+  $supervisor_service = $zulip::common::supervisor_service
+
   package { 'supervisor': ensure => 'installed' }
 
-  if $::osfamily == 'redhat' {
-    file { $zulip::common::supervisor_conf_dir:
-      ensure => 'directory',
-      owner  => 'root',
-      group  => 'root',
-    }
+  file { $zulip::common::supervisor_conf_dir:
+    ensure  => 'directory',
+    require => Package['supervisor'],
+    owner   => 'root',
+    group   => 'root',
+    purge   => true,
+    recurse => true,
+    notify  => Service[$supervisor_service],
   }
```

Comment:
One thing which isn't ideal about this change is that it makes it impossible for another service not managed by Zulip Puppet to be running inside supervisord.

Obviously we aren't currently prioritizing trying to make it easy for Zulip to share hosting with other services, but there are valid use cases for having custom supervisord config there (E.g. running the `zulip-botserver` service), and I'd like to avoid having Zulip's puppet rules aggressively purge configuration in directories that are not solely managed by Zulip.

One theory would be that we should use a separate `.d` directory for zulip-managed supervisord configuration that we can manage fully?  I'm not sure if supervisord has a good way to declare a second include directory, but we could look.

Reply:
> One thing which isn't ideal about this change is that it makes it impossible for another service not managed by Zulip Puppet to be running inside supervisord.

Yeah, I called that out in the commit message.

> One theory would be that we should use a separate .d directory for zulip-managed supervisord configuration that we can manage fully? I'm not sure if supervisord has a good way to declare a second include directory, but we could look.

Supervisor very much supports having multiple directories:
```
root@alexmv-prod:/etc/supervisor/conf.d# find /etc/supervisor/ -type f
/etc/supervisor/supervisord.conf
/etc/supervisor/conf.d/zulip/zulip_db.conf
/etc/supervisor/conf.d/zulip/zulip.conf
/etc/supervisor/conf.d/zulip/thumbor.conf
root@alexmv-prod:/etc/supervisor/conf.d# tail -2 /etc/supervisor/supervisord.conf
[include]
files = /etc/supervisor/conf.d/*.conf /etc/supervisor/conf.d/zulip/*.conf
root@alexmv-prod:/etc/supervisor/conf.d# supervisorctl status
process-fts-updates                              RUNNING   pid 592472, uptime 0:00:53
zulip-django                                     RUNNING   pid 592470, uptime 0:00:53
zulip-thumbor                                    RUNNING   pid 592473, uptime 0:00:53
zulip-tornado                                    RUNNING   pid 592471, uptime 0:00:53
zulip-workers:zulip_deliver_enqueued_emails      RUNNING   pid 592474, uptime 0:00:53
zulip-workers:zulip_deliver_scheduled_messages   RUNNING   pid 592476, uptime 0:00:53
zulip-workers:zulip_events                       RUNNING   pid 592475, uptime 0:00:53
```
But the problem is that getting into that state requires solving this very problem -- knowing which files from `/etc/supervisor/conf.d/` we need to remove because we've just installed a copy into `/etc/supervisor/conf.d/zulip/`.  The fact that they're installed into a separate directory may give us some leverage (on changes to `/etc/supervisor/conf.d/zulip/` exec something which removes same-name files from `/etc/supervisor/conf.d/`) but it's still somewhat brittle.

I'm sympathetic to wanting to run other services in supervisor on the host.  One counter-argument (which I'm not sure I'm convinced by 100% myself) is that those changes can just be local patches to puppet, which means that they're safe if the host needs to be reinstalled.

Reply:
The `/etc/supervisor/conf.d/zulip/` strategy seems like a great place to migrate to.  Can't we just do an `ensure => absent` for the <5 files that we used to manage in the top-level `/etc/supervisor/conf.d/`?  We'd have to not delete those couple lines for a year or two, but I think that'd be a safe path forward.

Reply:
Done!